auth/dfa.js

```diff
@@ -0,0 +1,605 @@
+    role = roles[j];
+    current = role.roleId[k];
+  }
+  var start = j;
+  while(j < n) {
+    role = roles[j];
+    var c = role.roleId[k];
+    // Here we go through the roles in the sorted order given, and whenever we
+    // encounter a new character we generate a DFA state for it, giving it the
+    // start and end-offset in the roles list of when we first saw the character
+    // and when it ended.
+    if (c !== current) {
+      state[current] = generateDFA(roles, start, j, k + 1, sets, implied);
+      current = c;
+      start = j;
+      splitCount += 1;
+    }
+    j += 1;
+  }
+  // Same as inside the loop, just for the last character
+  state[current] = generateDFA(roles, start, j, k + 1, sets, implied);
+
+  // There must always be a transition for the '*', if it is matched in the
+  // scope we're scanning then we must go to a state where if the scope ends
+  // you get all the roles is the current sub-tree. This is because '*' at the
+  // very end of a scope grants all the scopes matching it up to that '*'.
+  var star = state['*'] = state['*'] || {default: implied};
+
+  // if there is only one transition from this state then the current state has
+  // the same sub-tree as the state of that transition. Hence, we can just take
+  // sets index from the '*' -> end of that transition. Note, this is the place
+  // where we require that all states returned contains a '*' -> end transition.
+  if (splitCount === 0) {
+    star.end = state[current]['*'].end;
+  } else {
+    let set = roles.slice(afterImplied, n);
+    set.push(implied);
+    star.end = sets.push(set) - 1;
+  }
+
+  return state;
+};
+
+// Export generateDFA
+exports.generateDFA = generateDFA;
+
+/**
+ * Builds a pair {resolver, sets} where sets is a list of lists of roles,
+ * and given a scope `resolver(scope)`` returns an index from sets.
+ *
+ * That definition sounds slightly complicated, it's actually very simple,
+ * sets is on the form:
+ * ```js
+ * sets = [
+ *   [{role}, ...],
+ *   [{role}, ...],
+ *   [{role}, ...]
+ * ]
+ * ```
+ *
+ * For efficiency we allow sets[i] = [{role}, ..., j] where j < i to be
+ * interpreted as sets[i].concat(sets[j]). By not duplicating we don't have to
+ * resolve `sets[j]` multiple times, even though the set appears in multiple
+ * other sets.
+ *
+ * The `resolver` function returns an index in the sets array, as this allows
+ * us to later create a new sets variable where roles have been expanded to
+ * the scopes they imply, and just like that we can use the `resolver` to go
+ * from scope to expanded scopes.
+ */
+let buildResolver = (roles) => {
+  // Generate DFA
+  roles = sortRolesForDFAGeneration(roles);
+  let sets = [[]];
+  let dfa = generateDFA(roles, 0, roles.length, 0, sets, 0);
+
+  // Render a DFA state to code
+  let renderDFA = (state, depth) => {
+    var d = '';
+    while (d.length < depth * 4) d += '    ';
+    var c = '';
+    if (typeof(state.end) === 'number') {
```

Comment:
I think you're right, we always add `end` and nothing else `end == default`, but if in some future we break that variant and stop adding `end` when `end == default`, it'll be nice to have this condition.

But yeah, it's a nit.